lib/util/mp4_generator.js

```diff
@@ -0,0 +1,988 @@
+    let AVCLevelIndication = 0;
+    let profileCompatibility = 0;
+    for (let i = 0; i < this.videoNalus_.length; i++) {
+      const naluBytes = this.hexStringToBuffer_(this.videoNalus_[i]);
+      const naluType = naluBytes[0] & 0x1F;
+      switch (naluType) {
+        case NALUTYPE_SPS:
+          sps.push(naluBytes);
+          // 2 = sequenceParameterSetLength field length
+          avcCLength += naluBytes.length + 2;
+          break;
+        case NALUTYPE_PPS:
+          pps.push(naluBytes);
+          // 2 = pictureParameterSetLength field length
+          avcCLength += naluBytes.length + 2;
+          break;
+        default:
+          break;
+      }
+    }
+    // Get profile and level from SPS
+    if (sps.length > 0) {
+      AVCProfileIndication = sps[0][1];
+      profileCompatibility = sps[0][2];
+      AVCLevelIndication = sps[0][3];
+    }
+
+    // Generate avcC buffer
+    const avcCBytes = new Uint8Array(avcCLength);
+    let i = 0;
+    // configurationVersion = 1
+    avcCBytes[i++] = 1;
+    avcCBytes[i++] = AVCProfileIndication;
+    avcCBytes[i++] = profileCompatibility;
+    avcCBytes[i++] = AVCLevelIndication;
+    // '11111' + lengthSizeMinusOne = 3
+    avcCBytes[i++] = 0xFF;
+    // '111' + numOfSequenceParameterSets
+    avcCBytes[i++] = 0xE0 | sps.length;
+    for (let n = 0; n < sps.length; n++) {
+      avcCBytes[i++] = (sps[n].length & 0xFF00) >> 8;
+      avcCBytes[i++] = (sps[n].length & 0x00FF);
+      avcCBytes.set(sps[n], i);
+      i += sps[n].length;
+    }
+    // numOfPictureParameterSets
+    avcCBytes[i++] = pps.length;
+    for (let n = 0; n < pps.length; n++) {
+      avcCBytes[i++] = (pps[n].length & 0xFF00) >> 8;
+      avcCBytes[i++] = (pps[n].length & 0x00FF);
+      avcCBytes.set(pps[n], i);
+      i += pps[n].length;
+    }
+    const avcCBox = Mp4Generator.box('avcC', avcCBytes);
+
+    const avc1Bytes = new Uint8Array([
+      0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, // reserved
+      0x00, 0x01, // data_reference_index
+      0x00, 0x00, // pre_defined
+      0x00, 0x00, // reserved
+      0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00, // pre_defined
+      (width & 0xff00) >> 8,
+      width & 0xff, // width
+      (height & 0xff00) >> 8,
+      height & 0xff, // height
+      0x00, 0x48, 0x00, 0x00, // horizresolution
+      0x00, 0x48, 0x00, 0x00, // vertresolution
+      0x00, 0x00, 0x00, 0x00, // reserved
+      0x00, 0x01, // frame_count
+      0x13,
+      0x76, 0x69, 0x64, 0x65,
+      0x6f, 0x6a, 0x73, 0x2d,
+      0x63, 0x6f, 0x6e, 0x74,
+      0x72, 0x69, 0x62, 0x2d,
+      0x68, 0x6c, 0x73, 0x00,
+      0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, // compressorname
+      0x00, 0x18, // depth = 24
+      0x11, 0x11, // pre_defined = -1
+    ]);
+
+    const btrtBytes = new Uint8Array([
+      0x00, 0x1c, 0x9c, 0x80, // bufferSizeDB
+      0x00, 0x2d, 0xc6, 0xc0, // maxBitrate
```

Comment:
Are these values based on a spec or something?
If they're all just estimates... I think the `btrt` atom is optional, so it might just be better to not put it in.

Reply:
Removed the `btrt` atom